### PR TITLE
Enhanced support for Oracle SQL parsing.

### LINKED ISSUE #27069: Support parsing Oracle CREATE OR sql
# Background
Hi community,
This issue is for #26878.

ShardingSphere parser engine helps users parse a SQL to get the AST (Abstract Syntax Tree) and visit this tree to get SQLStatement (Java Object). Currently, we are planning to enhance the support for Oracle SQL parsing in ShardingSphere.

More details:
https://shardingsphere.apache.org/document/current/en/reference/sharding/parse/

# Task
This issue is to support more oracle sql parse, as follows:
```sql
CREATE OR REPLACE TRIGGER check_user
  AFTER LOGON ON DATABASE
  BEGIN
    check_user;
  EXCEPTION
    WHEN OTHERS THEN
      RAISE_APPLICATION_ERROR
        (-20000, 'Unexpected error: '|| DBMS_Utility.Format_Error_Stack);
 END;
```

```sql
CREATE OR REPLACE TRIGGER dept_cascade3
  AFTER UPDATE OF Deptno ON dept
BEGIN UPDATE emp
  SET Update_id = NULL
  WHERE Update_id = Integritypackage.Updateseq;
END;
```

```sql
CREATE OR REPLACE TRIGGER dept_del_cascade
  AFTER DELETE ON dept
  FOR EACH ROW

  -- Before row is deleted from dept,
  -- delete all rows from emp table whose DEPTNO is same as
  -- DEPTNO being deleted from dept table:

BEGIN
  DELETE FROM emp
  WHERE emp.Deptno = :OLD.Deptno;
END;
```

```sql
CREATE OR REPLACE TRIGGER dept_cascade2
  AFTER DELETE OR UPDATE OF Deptno ON dept
  FOR EACH ROW

  -- For each department number in dept that is updated,
  -- cascade update to dependent foreign keys in emp table.
  -- Cascade update only if child row was not updated by this trigger:
BEGIN
  IF UPDATING THEN
    UPDATE emp
    SET Deptno = :NEW.Deptno,
        Update_id = Integritypackage.Updateseq   --from 1st
    WHERE emp.Deptno = :OLD.Deptno
    AND Update_id IS NULL;

    /* Only NULL if not updated by 3rd trigger
       fired by same triggering statement */
  END IF;
  IF DELETING THEN
    -- After row is deleted from dept,
    -- delete all rows from emp table whose DEPTNO is same as
    -- DEPTNO being deleted from dept table:
    DELETE FROM emp
    WHERE emp.Deptno = :OLD.Deptno;
  END IF;
END;
```

```sql
CREATE OR REPLACE TRIGGER dept_restrict
  BEFORE DELETE OR UPDATE OF Deptno ON dept
  FOR EACH ROW

  -- Before row is deleted from dept or primary key (DEPTNO) of dept is updated,
  -- check for dependent foreign key values in emp;
  -- if any are found, roll back.

DECLARE
  Dummy                  INTEGER;  -- Use for cursor fetch
  Employees_present      EXCEPTION;
  employees_not_present  EXCEPTION;

  -- Cursor used to check for dependent foreign key values.
  CURSOR Dummy_cursor (Dn NUMBER) IS
    SELECT Deptno FROM emp WHERE Deptno = Dn;

BEGIN
  OPEN Dummy_cursor (:OLD.Deptno);
  FETCH Dummy_cursor INTO Dummy;

  -- If dependent foreign key is found, raise user-specified
  -- error code and message. If not found, close cursor
  -- before allowing triggering statement to complete.

  IF Dummy_cursor%FOUND THEN
    RAISE Employees_present;     -- Dependent rows exist
  ELSE
    RAISE Employees_not_present; -- No dependent rows exist
  END IF;
  CLOSE Dummy_cursor;

EXCEPTION
  WHEN Employees_present THEN
    CLOSE Dummy_cursor;
    Raise_application_error(-20001, 'Employees Present in'
      || ' Department ' || TO_CHAR(:OLD.DEPTNO));
  WHEN Employees_not_present THEN
    CLOSE Dummy_cursor;
END;
```

# Process
1. First confirm that this is a correct oracle sql syntax, if not please ignore;
2. Compare SQL definitions in Oficial SQL Doc and ShardingSphere SQL Doc;
3. If there is any difference in ShardingSphere SQL Doc, please correct them by referring to the Official SQL Doc;
4. Run mvn install the current_file_module;
5. Check whether there are any exceptions. If indeed, please fix them. (Especially xxxVisitor.class);
6. Add new corresponding SQL case in SQL Cases and expected parsed result in Expected Statment XML;
7. Run SQLParserParameterizedTest to make sure no exceptions.

# Relevant Skills
1. Master JAVA language
2. Have a basic understanding of Antlr `g4` file
3. Be familiar with Oracle SQLs

### PR DESCRIPTION
**Pull request by imzhangds**
Fixes #27069.

Changes proposed in this pull request:
  - Enhanced support for Oracle SQL parsing, fixed some issues with SQL parsing failure.
  - Added test cases for these SQL statements, and all tests passed successfully when running the test suite.
  
image

### Supported SQL cases are as follows
**case1**

```sql
CREATE OR REPLACE TRIGGER check_user
  AFTER LOGON ON DATABASE
  BEGIN
    check_user;
  EXCEPTION
    WHEN OTHERS THEN
      RAISE_APPLICATION_ERROR
        (-20000, 'Unexpected error: '|| DBMS_Utility.Format_Error_Stack);
 END;
```

**case2**

```sql
CREATE OR REPLACE TRIGGER dept_cascade3
  AFTER UPDATE OF Deptno ON dept
BEGIN UPDATE emp
  SET Update_id = NULL
  WHERE Update_id = Integritypackage.Updateseq;
END;
```

**case3**

```sql
CREATE OR REPLACE TRIGGER dept_del_cascade
  AFTER DELETE ON dept
  FOR EACH ROW

  -- Before row is deleted from dept,
  -- delete all rows from emp table whose DEPTNO is same as
  -- DEPTNO being deleted from dept table:

BEGIN
  DELETE FROM emp
  WHERE emp.Deptno = :OLD.Deptno;
END;
```

**case4**

```sql
CREATE OR REPLACE TRIGGER dept_cascade2
  AFTER DELETE OR UPDATE OF Deptno ON dept
  FOR EACH ROW

  -- For each department number in dept that is updated,
  -- cascade update to dependent foreign keys in emp table.
  -- Cascade update only if child row was not updated by this trigger:
BEGIN
  IF UPDATING THEN
    UPDATE emp
    SET Deptno = :NEW.Deptno,
        Update_id = Integritypackage.Updateseq   --from 1st
    WHERE emp.Deptno = :OLD.Deptno
    AND Update_id IS NULL;

    /* Only NULL if not updated by 3rd trigger
       fired by same triggering statement */
  END IF;
  IF DELETING THEN
    -- After row is deleted from dept,
    -- delete all rows from emp table whose DEPTNO is same as
    -- DEPTNO being deleted from dept table:
    DELETE FROM emp
    WHERE emp.Deptno = :OLD.Deptno;
  END IF;
END;
```

**case5**

```sql
CREATE OR REPLACE TRIGGER dept_restrict
  BEFORE DELETE OR UPDATE OF Deptno ON dept
  FOR EACH ROW

  -- Before row is deleted from dept or primary key (DEPTNO) of dept is updated,
  -- check for dependent foreign key values in emp;
  -- if any are found, roll back.

DECLARE
  Dummy                  INTEGER;  -- Use for cursor fetch
  Employees_present      EXCEPTION;
  employees_not_present  EXCEPTION;

  -- Cursor used to check for dependent foreign key values.
  CURSOR Dummy_cursor (Dn NUMBER) IS
    SELECT Deptno FROM emp WHERE Deptno = Dn;

BEGIN
  OPEN Dummy_cursor (:OLD.Deptno);
  FETCH Dummy_cursor INTO Dummy;

  -- If dependent foreign key is found, raise user-specified
  -- error code and message. If not found, close cursor
  -- before allowing triggering statement to complete.

  IF Dummy_cursor%FOUND THEN
    RAISE Employees_present;     -- Dependent rows exist
  ELSE
    RAISE Employees_not_present; -- No dependent rows exist
  END IF;
  CLOSE Dummy_cursor;

EXCEPTION
  WHEN Employees_present THEN
    CLOSE Dummy_cursor;
    Raise_application_error(-20001, 'Employees Present in'
      || ' Department ' || TO_CHAR(:OLD.DEPTNO));
  WHEN Employees_not_present THEN
    CLOSE Dummy_cursor;
END;
```

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by strongduanmu** [comment]
@imzhangds Nice to see you in the ShardingSphere.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 [hunk 1]
```diff
@@ -216,6 +217,12 @@ iterationCtlSeq
     : qualIterationCtl (COMMA_ qualIterationCtl)*
     ;
 
+modifyingExpression
+    : INSERTING
+    | DELETING
```
Comment (strongduanmu) [inline parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:222]: Can you combine these subrules into one line? Line breaks are required only when the subrule exceeds 5.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 [hunk 2]
```diff
@@ -314,6 +321,13 @@ ifStatement
     END IF SEMI_
     ;
 
+modifyingStatement
+    : IF modifyingExpression THEN statement+
```
Comment (strongduanmu) [inline parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:325]: Can you combine these subrules into one line? Line breaks are required only when the subrule exceeds 5.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 [hunk 3]
```diff
@@ -574,15 +588,15 @@ plsqlTriggerSource
     ;
 
 simpleDmlTrigger
-    : (BEFORE | AFTER) dmlEventClause (FOR EACH ROW)? triggerBody
+    : (BEFORE | AFTER) dmlEventClause (FOR EACH ROW)? () ? triggerBody
```
Comment (strongduanmu) [inline parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:591]: Is this writing legal in Oracle?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 2 above)
Comment (imzhangds) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:325]: Okay, I'll make the changes right away.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 3 above)
Comment (imzhangds) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:591]: Okay, I'll make the changes right away.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 1 above)
Comment (imzhangds) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:222]: Okay, I'll make the changes right away.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 2 above)
Comment (imzhangds) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4]: I have finished the modifications now.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 3 above)
Comment (imzhangds) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4]: I have finished the modifications now.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 1 above)
Comment (imzhangds) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4]: I have finished the modifications now.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 3 above)
Comment (strongduanmu) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4]: > I have finished the modifications now.

@imzhangds Good job. Do you want to get more deeply involved in the ShardingSphere? If you are interested, please contact me.

**Review by strongduanmu — APPROVED** [review]
LGTM, merged.

**Pull request merged**